### PR TITLE
*: replace all remaining `quick-error` with `thiserror`

### PR DESCRIPTION
**Pull request by koushiro**
### What problem does this PR solve?

Issue Number: last part of #8698, related issues: #9825 #9843 

### What is changed and how it works?

What's Changed: replace all remaining `quick-error` with `thiserror`

### Check List 

Tests 

- No code

Side effects

- No

### Release note 

No

### CODE REVIEW

**Comment by koushiro** [comment]
/cc @BusyJay @nrc 
Sorry for taking so long to complete the remaining part

**Comment by nrc** [comment]
> Sorry for taking so long to complete the remaining part

No problem, thanks for doing the work!

File: components/raftstore/src/errors.rs [hunk 1]
```diff
@@ -30,123 +30,91 @@ pub enum DiscardReason {
...
+        log_wrappers::Value::key(.1.get_end_key()),
+        .1.get_id()
+    )]
+    KeyNotInRegion(Vec<u8>, metapb::Region),
+    #[error("peer {1} is not ready, region {0} max_ts {2}")]
+    DataIsNotReady(
+        u64, // region_id
+        u64, // peer_id
+        u64, // safe_ts
+    ),
```
Comment (nrc) [inline components/raftstore/src/errors.rs:62]: Better to use a struct variant here

Comment (nrc) [reply components/raftstore/src/errors.rs:62]: Actually, I think that is true for most of these variants

File: src/storage/errors.rs [hunk 2]
```diff
@@ -5,88 +5,61 @@ use std::error;
...
+    Io(#[from] IoError),
+
+    #[error("scheduler is too busy")]
+    SchedTooBusy,
+
+    #[error("gc worker is too busy")]
+    GcWorkerTooBusy,
+
+    #[error("max key size exceeded, size: {0}, limit: {1}")]
+    KeyTooLarge(usize, usize),
```
Comment (nrc) [inline src/storage/errors.rs:49]: Should be a struct variant

File: components/raftstore/src/errors.rs (hunk 1 above)
Comment (koushiro) [reply components/raftstore/src/errors.rs]: Done

**Review by nrc — APPROVED** [review]
lgtm

**Review by andylokandy — APPROVED** [review]

**Comment by andylokandy** [comment]
@koushiro This test keeps falling

```
[2021-04-30T09:11:58.092Z] test cases::test_merge::test_node_merge_rollback ... thread 'main' panicked at 'response header { error { message: "1000 peer in merging mod, can't do proposal" proposal_in_merging_mode { region_id: 1000 } } current_term: 6 } has error', /home/jenkins/agent/tikv-master/build/components/test_raftstore/src/cluster.rs:924:13
[2021-04-30T09:11:58.092Z] stack backtrace:
[2021-04-30T09:11:58.092Z]    0: rust_begin_unwind
[2021-04-30T09:11:58.092Z]              at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/std/src/panicking.rs:493:5
[2021-04-30T09:11:58.092Z]    1: std::panicking::begin_panic_fmt
[2021-04-30T09:11:58.092Z]              at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/std/src/panicking.rs:435:5
[2021-04-30T09:11:58.092Z]    2: test_raftstore::cluster::Cluster<T>::must_put_cf
[2021-04-30T09:11:58.092Z]              at ./components/test_raftstore/src/cluster.rs:924:13
[2021-04-30T09:11:58.092Z]    3: test_raftstore::cluster::Cluster<T>::must_put
[2021-04-30T09:11:58.092Z]              at ./components/test_raftstore/src/cluster.rs:913:9
[2021-04-30T09:11:58.092Z]    4: failpoints::cases::test_merge::test_node_merge_rollback
[2021-04-30T09:11:58.092Z]              at ./tests/failpoints/cases/test_merge.rs:65:5
[2021-04-30T09:11:58.092Z]    5: failpoints::cases::test_merge::test_node_merge_rollback::{{closure}}
[2021-04-30T09:11:58.092Z]              at ./tests/failpoints/cases/test_merge.rs:25:1
[2021-04-30T09:11:58.092Z]    6: core::ops::function::FnOnce::call_once
[2021-04-30T09:11:58.092Z]              at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/ops/function.rs:227:5
[2021-04-30T09:11:58.092Z]    7: test_util::runner::run_test_with_hook::{{closure}}::{{closure}}
[2021-04-30T09:11:58.093Z]              at ./components/test_util/src/runner.rs:64:21
[2021-04-30T09:11:58.093Z]    8: core::ops::function::FnOnce::call_once{{vtable.shim}}
[2021-04-30T09:11:58.093Z]              at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/core/src/ops/function.rs:227:5
[2021-04-30T09:11:58.093Z]    9: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
[2021-04-30T09:11:58.093Z]              at /rustc/16bf626a31cb5b121d0bca2baa969b4f67eb0dab/library/alloc/src/boxed.rs:1546:9
[2021-04-30T09:11:58.093Z] note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
[2021-04-30T09:11:58.094Z] FAILED
```

**Comment by koushiro** [comment]
@andylokandy PTAL again

**Pull request merged**